docs/architecture/adr-versions.md

```diff
@@ -0,0 +1,219 @@
+
+P2PVersion is the first field in the NodeInfo. NodeInfo is also proto3 so this is easy to read out.
+
+For each P2PVersion, we keep a list of previous versions that this one is compatible with so we can connect to
+peers of any of those versions or higher.
+
+Note we need the peer/reactor protocols to take the versions of peers into account when sending messages:
+    - don't send messages they don't understand
+    - don't send messages they don't expect
+Doing this will be specific to the upgrades being made.
+
+Note we also include the list of reactor channels in the NodeInfo and already don't send messages for channels the peer doesn't understand.
+If upgrades always use new channels, this simplifies the development cost of backwards compatibility.
+
+Note NodeInfo is only exchanged after the authenticated encryption handshake to ensure that it's private.
+Doing any version exchange before encrypting could be considered information leakage, though I'm not sure
+how much that matters compared to being able to upgrade the protocol.
+
+XXX: if needed, can we change the meaning of the first byte of the first message to encode a handshake version?
+this is the first byte of a 32-byte ed25519 pubkey.
+
+#### AppVersion
+
+AppVersion is also included in the block Header and the NodeInfo.
+
+We only maintain connections to peers with the correct AppVersion for the height they're at.
+That is, we only connect to current peers with the same AppVersion, and to old peers that have the right AppVersion for where they
+are in the syncing process.
+
+Note this requires all consensus nodes to upgrade their versions at exactly the required height.
+Effectively doing so requires an external manager process that knows when the version should be updated and can co-ordinate shutting down,
+upgrading and restarting at the correct height.
+
+While this requires additional overhead in the form of a new manager process, it avoids the state machine logic getting
+cluttered with conditionals to effect different rules at different heights.
+
+More details on how this manager process works later below.
+
+### ChainDescription
+
+ChainDescription is a complete immutable description of a blockchain. It takes the following form:
+
+```
+ChainDescription = <NetworkName>/<BlockVersion>/<AppVersion>/<StateHash>/<ValHash>/<ConsensusParamsHash>
+```
+
+Here, StateHash is the merkle root of the initial state, ValHash is the merkle root of the initial Tendermint validator set,
+and ConsensusParamsHash is the merkle root of the initial Tendermint consensus parameters.
+
+The `genesis.json` file must contain enough information to compute this value. It need not contain the StateHash or ValHash itself,
+but contain the state from which they can be computed with the given protocol versions.
+
+XXX: should we split NetworkName into NetworkName and AppName?
+
+#### ChainID
+
+Define `ChainID = TMHASH(ChainDescriptor)`. It's the unique ID of a blockchain.
```

Comment:
What is the usage of chainID?

Reply:
It's the ultimate and final identifier for the chain. It's included in the SignBytes for votes and proposals so they are tied to only this chain.

Reply:
Should the ChainID be `TMHASH(ChainDescriptor || genesis state)`? Then this provides some degree of fork seperation. (Most forks would have a different genesis state, and therefore different chain id)

Reply:
ChainDescriptor already includes the state